unified-scheduler-pool/src/lib.rs

```diff
@@ -194,34 +199,535 @@ impl TaskHandler for DefaultTaskHandler {
+        let (accumulated_result_sender, accumulated_result_receiver) =
+            unbounded::<Option<ResultWithTimings>>();
+
+        let mut result_with_timings = self.session_result_with_timings.take();
+
+        // High-level flow of new tasks:
+        // 1. the replay stage thread send a new task.
+        // 2. the scheduler thread accepts the task.
+        // 3. the scheduler thread dispatches the task after proper locking.
+        // 4. the handler thread processes the dispatched task.
+        // 5. the handler thread reply back to the scheduler thread as an executed task.
+        // 6. the scheduler thread post-processes the executed task.
+        // 7. the scheduler thread send the executed task to the accumulator thread.
+        // 8. the accumulator thread examines the executed task's result and accumulate its timing,
+        //    finally dropping the transaction inside the executed task.
+        let scheduler_main_loop = || {
+            let handler_count = self.handler_count;
+            let session_result_sender = self.session_result_sender.clone();
+            let new_task_receiver = self.new_task_receiver.clone();
+
+            let mut session_ending = false;
+            let mut active_task_count: usize = 0;
+
+            // Now, this is the main loop for the scheduler thread, which is a special beast.
+            //
+            // That's because it's the most notable bottleneck of throughput. Unified scheduler's
+            // overall throughput is largely dependant on its ultra-low latency characteristic,
+            // which is the most important design goal of the scheduler in order to reduce the
+            // transaction confirmation latency for end users.
+            //
+            // Firstly, the scheduler thread must handle incoming messages from thread(s) owned by
+            // the replay stage or the banking stage. It also must handle incoming messages from
+            // the multi-threaded handlers. This heavily-multi-threaded whole processing load must
+            // be coped just with the single-threaded scheduler, to attain ideal cpu cache
+            // friendliness and main memory bandwidth saturation with its shared-nothing
+            // single-threaded account locking implementation. In other words, the per-task
+            // processing efficiency of the main loop codifies the upper bound of horizontal
+            // scalability of the unified scheduler.
+            //
+            // Moreover, the scheduler is designed to handle tasks without batching at all in the
+            // pursuit of saturating all of the handler threads with maximally-fine-grained
+            // concurrency density for throughput as the second design goal. This design goal
+            // relies on the assumption that there's no considerable penalty arising from the
+            // unbatched manner of processing.
```

Comment:
To be clear, I'm not strictly opposed to eliminating batching in replay.
It's probable that the gained parallelism from not doing batches can, in many cases, make it overall more efficient.

But I think we shouldn't state the assumption there's no significant loss, when the data suggests otherwise.

Reply:
> 20us/tx => 1280us/(64tx). So there's pretty significant loss of efficiency by eliminating batching entirely.

well, i think i managed to eliminate the overhead: https://github.com/apfitzge/solana/pull/3/files

tested on old intel cpu

```
running 8 tests
test bench_execute_batch_01_txes_per_batch          ... bench:     199,425 ns/iter (+/- 21,345)
test bench_execute_batch_04_txes_per_batch          ... bench:     203,539 ns/iter (+/- 31,947)
test bench_execute_batch_16_txes_per_batch          ... bench:     211,507 ns/iter (+/- 72,429)
test bench_execute_batch_64_txes_per_batch          ... bench:     180,725 ns/iter (+/- 36,828)
test bench_execute_batch_flattend_01_txes_per_batch ... bench:     182,158 ns/iter (+/- 25,963)
test bench_execute_batch_flattend_04_txes_per_batch ... bench:     288,362 ns/iter (+/- 13,242)
test bench_execute_batch_flattend_16_txes_per_batch ... bench:     181,084 ns/iter (+/- 17,782)
test bench_execute_batch_flattend_64_txes_per_batch ... bench:     185,192 ns/iter (+/- 26,854)
```

tested on recent amd cpu:

```
running 8 tests
test bench_execute_batch_01_txes_per_batch          ... bench:     210,430 ns/iter (+/- 4,364)
test bench_execute_batch_04_txes_per_batch          ... bench:     177,109 ns/iter (+/- 5,016)
test bench_execute_batch_16_txes_per_batch          ... bench:     188,921 ns/iter (+/- 7,451)
test bench_execute_batch_64_txes_per_batch          ... bench:     180,847 ns/iter (+/- 5,388)
test bench_execute_batch_flattend_01_txes_per_batch ... bench:     183,689 ns/iter (+/- 4,730)
test bench_execute_batch_flattend_04_txes_per_batch ... bench:     183,313 ns/iter (+/- 7,748)
test bench_execute_batch_flattend_16_txes_per_batch ... bench:     176,748 ns/iter (+/- 1,863)
test bench_execute_batch_flattend_64_txes_per_batch ... bench:     177,591 ns/iter (+/- 4,527)

```

Reply:
in that dirtiest branch, i disabled most of non-essential code, which is causing those overheads. i think all of them are fixable to have no overhead with moderate amount efforts.

indeed, unbatched code path incurs raw increase of instruction count. however, these are basically can be offset for better data/code locality. i can't write in detail (it's midnight here..). but i confirmed this with amd/intel with `perf stat`.

so, this is a glimpse of what I'm arguing against batching for years. :)